### ISSUE TITLE
Inconsistent results from `mask == 0`

### ISSUE THREAD

**Issue by zhangguanheng66** [pytorch/pytorch#28508]
## 🐛 Bug

With the master branch or v.1.3, the results of masked_fill func are not correct. There may be some changes that I don't realize.

Because of this issue, my transform example is broken (https://github.com/pytorch/examples/issues/642).
 

## To Reproduce

1. Use the follow code to reproduce the error.
```
import torch
print(torch.__version__)
mask = (torch.triu(torch.ones(3, 3)) == 1).transpose(0, 1)
print(mask)
mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
print(mask)
```

With PyTorch 1.3 or above, I have
```
1.4.0a0+3fce612
tensor([[ True, False, False],
        [ True,  True, False],
        [ True,  True,  True]])
tensor([[0., 0., 0.],
        [-inf, 0., 0.],
        [-inf, -inf, 0.]])
```

## Expected behavior

With PyTorch 1.2, I have
```
1.2.0a0+54a63e0
tensor([[ True, False, False],
        [ True,  True, False],
        [ True,  True,  True]])
tensor([[0., -inf, -inf],
        [0., 0., -inf],
        [0., 0., 0.]])
```

## Environment

PyTorch version: 1.4.0a0+3fce612
Is debug build: No
CUDA used to build PyTorch: 9.2

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 410.79
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.4
[pip] numpydoc==0.9.1
[pip] torch==1.4.0a0+3fce612
[pip] torchtext==0.4.0
[conda] blas                      1.0                         mkl
[conda] cuda92                    1.0                           0    pytorch
[conda] magma-cuda92              2.5.1                         1    pytorch
[conda] mkl                       2019.4                      243
[conda] mkl-include               2019.4                      243
[conda] mkl-service               2.0.2            py37h7b6447c_0
[conda] mkl_fft                   1.0.12           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch                   1.2.0           py3.7_cuda10.0.130_cudnn7.6.2_0    pytorch
[conda] torch                     1.4.0a0+3fce612           dev_0    <develop>

## Additional context

cc @ezyang @gchanan @zou3519 @jerryzh168

**Comment by zhangguanheng66** [comment]
@zou3519 @izdeby any ideas?

**Comment by izdeby** [comment]
I'm not sure what might it be. Looks like a regression. Would have to take a look to figure out

**Comment by zhangguanheng66** [comment]
High priority because it broke nn.Transformer examples and tutorial. Better to fix it soon.

**Comment by albanD** [comment]
Hi,

I think this is not due to masked_fill but some error in type promotion, see https://github.com/pytorch/pytorch/issues/28502

Could you try the same code with `mask == False` instead of `mask == 0` ? Does that solve the issue?

**Comment by zhangguanheng66** [comment]
Yes. I had a similar thought. Switching to `mask == False` gives me consistent results.

**Comment by albanD** [comment]
I'll close this in favor of the original issue then.
Thanks for the report !

**Comment by zhangguanheng66** [comment]
I want to keep it open for now since it's high priority issue. Both transformer tutorial and examples are affected.

**Comment by albanD** [comment]
Sure, could you just rename it to state that the issue is with the `mask == 0` line and has nothing to do with the masked_fill function?

**Comment by nairbv** [comment]
this is fixed by:
https://github.com/pytorch/pytorch/pull/28253
